Add Standard.Tableau to native-image

hubertp opened the pull request:
### Pull Request Description

Closes #12194.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Add Standard.Tableau to native-image

Closes #12194.

hubertp committed:
Fix sbt warning

```
[warn] there's a key that's not used by any other settings/tasks:
[warn]
[warn] * runtime-compiler-dump / javaModuleName
[warn]   +- /home/hubert/tmp/enso/build.sbt:3425
[warn]
[warn] note: a setting might still be used by a command; to exclude a key from this `lintUnused` check
[warn] either append it to `Global / excludeLintKeys` or call .withRank(KeyRanks.Invisible) on the key
```

hubertp committed:
Test NI in nightlies

JaroslavTulach commented:
[70MB in resources](https://github.com/enso-org/enso/actions/runs/13072404716/job/36476784788?pr=12204#step:8:2128)!? Are we distributing another `.so` or `.dll` library?

JaroslavTulach commented on std-bits/tableau/src/main/resources/META-INF/native-image/org/enso/tableau/resource-config.json:
```diff
@@ -0,0 +1,22 @@
+                "pattern": "\\QMETA-INF/services/com.tableau.hyperapi.impl.SharedLibraryProvider\\E"
+            },
+            {
+                "pattern": "\\Qlinux-x86-64/libtableauhyperapi.so\\E"
```
These libraries should be extracted into `polyglot/lib` directory.

hubertp replied:
At one time there is only a single OS-specific jar with the library. `resource-config` is simply not aware of OS/Arch so listing them here is of little harm, I believe. 

I haven't changed in any way the way we package/include shared library and os/arch-specific jar has always been included in `polyglot`. What do we gain by this proposal and why this seems to be NI-specific to you and should be addressed in this PR?

hubertp replied:
Unless... placing library in `polyglot/lib` means that it will get automatically included. Then yes, it totally makes sense!

JaroslavTulach replied:
- @Akirathan worked on #11874 
- making the `polyglot/lib` an easy to use _standard way of packaging JNI libraries_
    - makes the distribution smaller: #11483
    - makes the startup faster (no extraction of a resource on start)
    - avoids duplicating the library between JAR and _native executable_
- the whole system was successfully used in #12021
- we should use it for all our JNI libraries (tableau, postgres, sqlite, snowflake)
- unless there is a reason not to

hubertp replied:
Yeap, totally makes sense, I somehow forgot about it. Thanks for bringing it up.

hubertp replied:
This was a bit more messy than initially expected. Had to provide a custom classloader logic that intercepts hyper api search that is done in Tableau's library.

JaroslavTulach reviewed (changes requested):
`tableauhperap`i library should be extracted from the JAR into `polyglot/lib` directory.

hubertp committed:
Rewrite library loading for NI's Tableau

This change extracts `Tableau`'s Hyper shared library from the jar.
Unlike `Standard.Image`, had to provide a custom classloader that
manages to find requested library - HyperAPI used a custom JNA logic for
that.

hubertp committed:
Merge branch 'develop' into wip/hubert/12194-ni-tableau

hubertp committed:
fmt

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java:
```diff
@@ -172,6 +175,20 @@ private static Module registerModule(
+          return libPath;
+        }
+      }
+      throw new UnsupportedOperationException("unable to find library " + libname);
```
Note: I'm not entirely sure this is the right way but `RuntimeException` didn't feel right either. 
Maybe a `PanicException` or a simple `null`?

Open for suggestions.

Akirathan replied:
This method is called only from `HyperReader`. I suggested implementing finding native library this way, because `HyperReader` is in `std-bits` and, therefore, has no (direct) access to `EnsoContext` and its `NativeLibraryFinder`. I don't have a strong opinion about the exception. But I suggest you append additional test for it inside [NativeLibraryFinderTest](https://github.com/enso-org/enso/blob/develop/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/NativeLibraryFinderTest.java).

JaroslavTulach replied:
If we are exposing this for the Enso code, then:
- `ctx.nothing()` might be a reasonable value. 
- `PanicException` is a good alternative as well
- [UnsupportedMessageException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedMessageException.html) is a poor's man solution as well

Certainly not `java.lang.UnsupportedOperationException` - that would mean crash of the interpreter!

hubertp commented:
> [70MB in resources](https://github.com/enso-org/enso/actions/runs/13072404716/job/36476784788?pr=12204#step:8:2128)!? Are we distributing another `.so` or `.dll` library?

Yes.

hubertp committed:
fmt

Akirathan commented on engine/common/src/main/java/org/enso/common/MethodNames.java:
```diff
@@ -9,6 +9,7 @@ public static class TopScope {
     public static final String REGISTER_MODULE = "register_module";
     public static final String UNREGISTER_MODULE = "unregister_module";
     public static final String COMPILE = "compile";
+    public static final String FIND_LIBRARY = "find_library";
```
Probably better to rename this to `FIND_NATIVE_LIBRARY`.

JaroslavTulach commented on std-bits/tableau/src/main/java/org/enso/tableau/HyperReader.java:
```diff
@@ -104,6 +107,30 @@ private static HyperProcess getProcess() throws IOException {
+    }
+
+    @Override
+    public InputStream getResourceAsStream(String name) {
```
What is the caller stack trace? What the tableu code is trying to achieve? Why does it need `InputStream`? Can we prevent it to ask for input stream?

hubertp replied:
We can't prevent it. It's part of hyper initialization.

JaroslavTulach replied:
My [investigation](https://github.com/enso-org/enso/pull/12204#discussion_r1940612856) suggests there is a way to prevent the call and skip the _"extraction of the JNI library"_ completely.

JaroslavTulach commented on std-bits/tableau/src/main/java/org/enso/tableau/HyperReader.java:
```diff
@@ -104,6 +107,30 @@ private static HyperProcess getProcess() throws IOException {
+      int dotIdx = name.indexOf(".");
+      if (libIdx != -1 && dotIdx != -1) {
+        var libName = name.substring(libIdx + 3, dotIdx);
+        var bindings = Context.getCurrent().getBindings("enso");
```
Calling `org.graalvm.polyglot` from `ClassLoader` - that's wild.

Akirathan replied:
The only other option how to get reference to the `NativeLibraryFinder` functionality is to move the `NativeLibraryFinder` out of `org.enso.runtime` into some module that is shared with std-bits. The approach via `Context.getCurrent().getBindings("enso")` is what I have suggested to @hubertp as the "best" solution that I can think of. I am open to other suggestions of course.

hubertp replied:
Care to expand on what's so _wild_ about it?

JaroslavTulach replied:
> what's so wild about it?

`ClassLoader` is a low level JVM mechanism. Calling from such a level into a Truffle language implemented on top of Truffle framework written in Java is just _wild_. It's like calling from a Linux kernel driver into X11 paint method and waiting for a result.

Doing something like this is a request to get a deadlock in internal JVM classloading locks.

hubertp replied:
#yolo

Jokes aside, it would be best to suggest some improvements upstream as those jars are not integration-friendly. But that can take _a lot_ of time and has unknown chances of success.

JaroslavTulach replied:
To avoid this _"kernel to graphics"_ dependency, I'd like to point your attention to [NetBeans translate support](https://bits.netbeans.org/dev/javadoc/org-openide-util/org/openide/util/BaseUtilities.html#translate-java.lang.String-):
- it provides a way to instruct the classloader to do something else
- by having a special file that provides a mapping
- all NetBeans `ClassLoader`s support this `translating`

We could let the Enso `tableau.jar` provide such a file to instruct the `HostClassLoader` to return the `FileInputStream` from `getResourceAsStream` when a particular resource cannot be found.

The difference to current state is that we don't need to call into _polyglot context_ from a `ClassLoading` code.

JaroslavTulach commented on std-bits/tableau/src/main/java/org/enso/tableau/HyperReader.java:
```diff
@@ -88,7 +91,7 @@ private static HyperProcess getProcess() throws IOException {
       var contextClassLoader = Thread.currentThread().getContextClassLoader();
       try {
-        Thread.currentThread().setContextClassLoader(HyperReader.class.getClassLoader());
+        Thread.currentThread().setContextClassLoader(new TableauClassLoader());
```
There seems to be:
```bash
$ javap -cp built-distribution/enso-engine-*/enso-*/lib/Standard/Tableau/0.0.0-dev/polyglot/java/tableauhyperapi.jar com.tableau.hyperapi.impl.SharedLibraryProvider
Compiled from "SharedLibraryProvider.java"
public interface com.tableau.hyperapi.impl.SharedLibraryProvider {
  public abstract java.lang.String getSharedLibraryBaseName();
}
```
cannot it be used for loading the JNI library without these classloading tricks? Looks like the code checks `jna.nounpack` property:
```
  private static void registerJNA();
    Code:
       0: ldc           #52                 // String jna.nounpack
       2: invokestatic  #53                 // Method java/lang/Boolean.getBoolean:(Ljava/lang/String;)Z
       5: ifeq          17
       8: invokestatic  #54                 // Method getLibraryBaseName:()Ljava/lang/String;
      11: invokestatic  #55                 // Method com/sun/jna/Native.register:(Ljava/lang/String;)V
```
With a proper value provided, the code should be able to skip this extracting step completely.

hubertp replied:
Yes, I saw that. Sadly I couldn't make it to work that way; JNA gets in the way.

JaroslavTulach replied:
OK, thanks.

hubertp committed:
Add tests

hubertp committed:
Provide more info to CI failures

JaroslavTulach reviewed (approved):
Let me approve in the _name of pragmatism_. Since my [previous review comment](https://github.com/enso-org/enso/pull/12204#discussion_r1938193643) this PR managed to:

> - make the distribution smaller:
> - avoid duplicating the library between JAR and native executable

Only the 

> - makes the startup faster (no extraction of a resource on start)

remains open. That's probably good enough for now - especially when it is unlikely many people use Tableau to begin with.

I still don't like the ["kernel to graphics call"](https://github.com/enso-org/enso/pull/12204#discussion_r1940946349), but I can live with it. Moreover I am going to make another proposal to solve it and [stay in the kernel space](https://github.com/enso-org/enso/pull/12204#discussion_r1941170133).

hubertp committed:
Extract native library from JNA

Seems to be causing problems on CI.
Extraction logic had to be extended to allow for:
- different prefixes (e.g, `linux-86-64` instead of `linux-86_64)
- classes being present on the prefix should be included in the thin
  jar; we cannot assume that extraction prefix is unique

hubertp committed:
Fix Windows and MacOS builds

hubertp committed:
Having a single name for OS would be too simple

hubertp committed:
One more attempt to make Windows happy

hubertp committed:
Set JNA boot dir explicitly

Alternative approach to deal with Windows/Mac failures on CI where
`jnidispatch` native library somehow could not be identified. Linux
worked just fine with the previous approach.

This change enhances custom classloader to identify the desired library.
That way we can provide the path to it, as specified in the
implementation of JNA.

JaroslavTulach commented:
_"Squash and merge"_ button is enabled. I guess this is the [failure holding us](https://github.com/enso-org/enso/actions/runs/13168597968/job/36754393426?pr=12204#step:8:5877) from pushing it. I guess I can see the same failure on my Intel Mac:
```
[error] Test org.enso.interpreter.test.NativeLibraryFinderTest.standardTableauShouldHaveNativeLib failed: java.lang.AssertionError: 
[error]   os.name: Mac OS X
[error]   os.arch: x86_64
[error]   Mapped library name: 
[error]       libopencv_java470.dylib
[error]       libtableauhyperapi.dylib
[error]   Contents of Standard.Tableau native library dir:
[error]    /enso/distribution/lib/Standard/Tableau/0.0.0-dev/polyglot/lib/amd64/libtableauhyperapi.dylib
[error] , took 4.014 sec
[error]     at org.enso.interpreter.test.NativeLibraryFinderTest$PrintSystemInfoRule$1.evaluate(NativeLibraryFinderTest.java:164)
[error]     ...
[error] Caused by: java.lang.AssertionError: There should be two native libs for Standard.Tableau
[error] Expected: is <2>
[error]      but: was <1>
[error]     at org.enso.interpreter.test.NativeLibraryFinderTest.standardTableauShouldHaveNativeLib(NativeLibraryFinderTest.java:80)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
[error]     ...
[info] Test run org.enso.interpreter.test.NativeLibraryFinderTest finished: 1 failed, 0 ignored, 4 total, 21.906s
[error] Failed: Total 4, Failed 1, Errors 0, Passed 3
[error] Failed tests:
[error]         org.enso.interpreter.test.NativeLibraryFinderTest
```
btw. I can execute the same test on my M2 Mac and it fails too, but differently:
```
sbt:enso> runtime-integration-tests/testOnly *NativeLibraryFinderTest
[error] Test org.enso.interpreter.test.NativeLibraryFinderTest.standardImageShouldHaveNativeLib failed: java.lang.AssertionError: 
[error]   os.name: Mac OS X
[error]   os.arch: aarch64
[error]   Mapped library name: 
[error]       libopencv_java470.dylib
[error]       libtableauhyperapi.dylib
[error]   Contents of Standard.Image native library dir:
[error] , took 1.163 sec
[error]     at org.enso.interpreter.test.NativeLibraryFinderTest$PrintSystemInfoRule$1.evaluate(NativeLibraryFinderTest.java:164)
[error]     ...
[error] Caused by: java.lang.AssertionError: There should be just single native lib in Standard.Image
[error] Expected: is <1>
[error]      but: was <0>
[error]     at org.enso.interpreter.test.NativeLibraryFinderTest.standardImageShouldHaveNativeLib(NativeLibraryFinderTest.java:56)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
[error]     ...
[error] Test org.enso.interpreter.test.NativeLibraryFinderTest.standardTableauShouldHaveNativeLib failed: java.lang.AssertionError: 
[error]   os.name: Mac OS X
[error]   os.arch: aarch64
[error]   Mapped library name: 
[error]       libopencv_java470.dylib
[error]       libtableauhyperapi.dylib
[error]   Contents of Standard.Tableau native library dir:
[error]     /enso/distribution/lib/Standard/Tableau/0.0.0-dev/polyglot/lib/aarch64/libtableauhyperapi.dylib
[error] , took 1.489 sec
[error]     at org.enso.interpreter.test.NativeLibraryFinderTest$PrintSystemInfoRule$1.evaluate(NativeLibraryFinderTest.java:164)
[error]     ...
[error] Caused by: java.lang.AssertionError: There should be two native libs for Standard.Tableau
[error] Expected: is <2>
[error]      but: was <1>
[error]     at org.enso.interpreter.test.NativeLibraryFinderTest.standardTableauShouldHaveNativeLib(NativeLibraryFinderTest.java:80)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
[error]     ...
[info] Test run org.enso.interpreter.test.NativeLibraryFinderTest finished: 2 failed, 0 ignored, 4 total, 8.321s
[error] Failed: Total 4, Failed 2, Errors 0, Passed 2
[error] Failed tests:
[error]         org.enso.interpreter.test.NativeLibraryFinderTest
[error] (runtime-integration-tests / Test / testOnly) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 24 s, completed 6. 2. 2025 6:37:29
```
on M2 Mac running the tableau tests in _native mode_ fails:
```
enso$ ENSO_LAUNCHER=native,test,fast,-ls sbt 'runEngineDistribution --run test/Tableau_Tests'

3 tests succeeded.
10 tests failed.
0 tests skipped.
0 groups skipped.
```
If you need me to run or investigate something else on my Macs, I can...

hubertp commented:
Thanks. I solved Windows yesterday but looks like Mac has its own set of issues. Let me take another look.

hubertp committed:
JNA uses deprecated extensions for native libs

`s/jnilib/dylib`

hubertp commented:
> Thanks. I solved Windows yesterday but looks like Mac has its own set of issues. Let me take another look.

I think that one is resolved as well, just CI/GH complaining about random stuff.

hubertp committed:
Merge branch 'develop' into wip/hubert/12194-ni-tableau

hubertp committed:
Partial revert of #12197

Breaks Tableau integration to accomadate for opencv non-standard
packaging.

hubertp committed:
Merge branch 'develop' into wip/hubert/12194-ni-tableau

JaroslavTulach commented on build.sbt:
```diff
@@ -5147,15 +5162,30 @@ lazy val `std-tableau` = project
+          `std-tableau-polyglot-root`,
+          `std-tableau-native-libs`,
+          tableauVersion,
+          jnaVersion
```
- for the purpose of https://github.com/enso-org/enso/pull/12843#issuecomment-2846299528 
- I am trying to patch JNA
- it would help, if there was just a **single copy** of the JNA classes in `built-distribution`

JaroslavTulach commented on project/StdBits.scala:
```diff
@@ -166,18 +167,106 @@ object StdBits {
+      streams.value.cacheStoreFactory
+    )
+
+    // Extract native library from jna's jar
```
- if there has to be a copy of the JNA JAR file
- I'd like to take it from `jni-wrapper` JAR
- and not directly from `net.java.dev.jna:jna` Maven dependency

JaroslavTulach reviewed (commented):
Wrote a small shell script to search for JNA classes:
```bash
#!/bin/bash
function findJars() {
  find built-distribution/ | grep jar$ | while read X; do (unzip -v $X | grep com.sun.jna.Native) && echo JNA:$X; done
}

findJars | grep ^JNA: | cut -f2 -d : | while read X; do
    echo Checking $X
    javap -private -cp $X com.sun.jna.Native | grep java.awt
done
```
right now (e.g. `2025.2.1-nightly.2025.5.2`) it finds too many copies of JNA:
```
Checking built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Tableau/0.0.0-dev/polyglot/java/jna-5.14.0.jar
  public static long getWindowID(java.awt.Window) throws java.awt.HeadlessException;
  public static long getComponentID(java.awt.Component) throws java.awt.HeadlessException;
  public static com.sun.jna.Pointer getWindowPointer(java.awt.Window) throws java.awt.HeadlessException;
  public static com.sun.jna.Pointer getComponentPointer(java.awt.Component) throws java.awt.HeadlessException;
  static native long getWindowHandle0(java.awt.Component);
Checking built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Tableau/0.0.0-dev/polyglot/java/jna-linux-5.14.0.jar
  public static long getWindowID(java.awt.Window) throws java.awt.HeadlessException;
  public static long getComponentID(java.awt.Component) throws java.awt.HeadlessException;
  public static com.sun.jna.Pointer getWindowPointer(java.awt.Window) throws java.awt.HeadlessException;
  public static com.sun.jna.Pointer getComponentPointer(java.awt.Component) throws java.awt.HeadlessException;
  static native long getWindowHandle0(java.awt.Component);
Checking built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/jna-wrapper-assembly-0.1.0-SNAPSHOT.jar
  public static long getWindowID(java.awt.Window) throws java.awt.HeadlessException;
  public static long getComponentID(java.awt.Component) throws java.awt.HeadlessException;
  public static com.sun.jna.Pointer getWindowPointer(java.awt.Window) throws java.awt.HeadlessException;
  public static com.sun.jna.Pointer getComponentPointer(java.awt.Component) throws java.awt.HeadlessException;
  static native long getWindowHandle0(java.awt.Component);
```
just FYI. No changes in this PR (of course) needed.
